synapse/storage/databases/main/relations.py

```diff
@@ -91,10 +91,18 @@ def __init__(
 
         self._msc3440_enabled = hs.config.experimental.msc3440_enabled
 
+    def _invalidate_relations_for_event(self, event_id: str) -> None:
+        """Invalidate the relation caches for an event."""
+        self.get_relations_for_event.invalidate((event_id,))
+        self.get_aggregation_groups_for_event.invalidate((event_id,))
+        self.get_applicable_edit.invalidate((event_id,))
+        self.get_thread_summary.invalidate((event_id,))
+        self.get_thread_participated.invalidate((event_id,))
+
     @cached(tree=True)
     async def get_relations_for_event(
         self,
-        event_id: str,
+        event: EventBase,
```

Comment:
I wonder if this messes up the caching...

Reply:
See https://github.com/matrix-org/synapse/pull/11915#discussion_r818657212 for more info...

Reply:
Hmm -- I'm struggling to refactor this a bit since we would *have* to provide the `event_id` (and a matching `event`), but a bunch of the other properties, which we cache on, are optional / have defaults.

Any thoughts on how to refactor this? Or maybe we could have `cached` learn that an `EventBase` should be cached by event ID instead of the actual object?

Reply:
> Any thoughts on how to refactor this? Or maybe we could have `cached` learn that an `EventBase` should be cached by event ID instead of the actual object?

I took a look at this and it seems doable, but probably adds code to a hot path which would be used rarely. As far as I can tell there's only 3 situations where we do something similar:

1. `get_relations_for_event` and `is_interested_in_event` (over `EventBase` and its `event_id` property).
2. `_get_joined_hosts` (over `_StateCacheEntry` and its `state_group` property).

This one is a bit ickier though since we have other parameters after it, I think the best thing to do is just default it to `None`, then raise an error if it is `None`.